flink-ml-core/src/main/java/org/apache/flink/ml/common/sharedstorage/SharedStorage.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    /** Gets a {@link Writer} of shared item identified by (storageID, subtaskId, key). */
+    static <T> Writer<T> getWriter(
+            StorageID storageID,
+            int subtaskId,
+            ItemDescriptor<T> descriptor,
+            String ownerId,
+            OperatorID operatorID,
+            StreamTask<?, ?> containingTask,
+            StreamingRuntimeContext runtimeContext,
+            StateInitializationContext stateInitializationContext) {
+        Tuple3<StorageID, Integer, String> t = Tuple3.of(storageID, subtaskId, descriptor.key);
+        String lastOwner = owners.putIfAbsent(t, ownerId);
+        if (null != lastOwner) {
+            throw new IllegalStateException(
+                    String.format(
+                            "The shared item (%s, %s, %s) already has a writer %s.",
+                            storageID, subtaskId, descriptor.key, ownerId));
+        }
+        Writer<T> writer =
+                new Writer<>(
+                        t,
+                        ownerId,
+                        descriptor.serializer,
+                        containingTask,
+                        runtimeContext,
+                        stateInitializationContext,
+                        operatorID);
+        writer.set(descriptor.initVal);
+        return writer;
+    }
+
+    static class Reader<T> {
+        protected final Tuple3<StorageID, Integer, String> t;
+
+        Reader(Tuple3<StorageID, Integer, String> t) {
+            this.t = t;
+        }
+
+        T get() {
+            // It is possible that the `get` request of an item is triggered earlier than its
+            // initialization. In this case, we wait for a while.
+            long waitTime = 10;
+            do {
+                //noinspection unchecked
+                T value = (T) m.get(t);
+                if (null != value) {
+                    return value;
+                }
+                try {
+                    Thread.sleep(waitTime);
+                } catch (InterruptedException e) {
+                    break;
+                }
+                waitTime *= 2;
+            } while (waitTime < 10 * 1000);
+            throw new IllegalStateException(
+                    String.format("Failed to get value of %s after waiting %d ms.", t, waitTime));
+        }
+    }
+
+    static class Writer<T> extends Reader<T> {
+        private final String ownerId;
+        private final ListStateWithCache<T> cache;
+        private boolean isDirty;
+
+        Writer(
+                Tuple3<StorageID, Integer, String> t,
+                String ownerId,
+                TypeSerializer<T> serializer,
+                StreamTask<?, ?> containingTask,
+                StreamingRuntimeContext runtimeContext,
+                StateInitializationContext stateInitializationContext,
+                OperatorID operatorID) {
+            super(t);
+            this.ownerId = ownerId;
+            try {
+                cache =
```

Comment:
I'm not quite sure what you mean by your suggestion. Are you suggesting to put the in-memory value in the `Writer` instance to avoid the lock?  

Following the previous reply, every subtask owns a distinct instance of `SharedStorageContextImpl`. So, the `Writer` instance stored in `SharedStorageContextImpl` is only visible to its owner subtask of one operator. It cannot be accessed by subtasks of other operators.